### PR TITLE
Simplify wild card matching

### PR DESCRIPTION
**Pull request by awemany**
- replace match(..) with libc function fnmatch(..)
- always use fnmatch instead of separate match for no wildcards

### CODE REVIEW

**Comment by sickpig** [comment]
https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/237150670#L2021
patch failed to cross compile for win32/64

**Comment by awemany** [comment]
Damn. There doesn't seem to be fnmatch for mingw/windows. Anyone familiar enough with Windows know a good replacement?

SO suggests PathMatchSpec(), but then we'd need to make some ugly #ifdef logic for that.

Should I look for a acceptably-licensed fnmatch(..) implementation to include in the source?

I rather prefer to use well-tested functions in well-known libraries.

I guess I also need help here by someone using Windows regularly who could test this.

@thezerg: What would be your preferred way forward here?

**Comment by sickpig** [comment]
maybe @ptschip and @Justaphf could have an idea on how to deal with lack of fnmatch on mingw.

**Comment by Justaphf** [comment]
@awemany I am not familiar with the Windows equivalent of `fnmatch`, but a quick Google search indicated the same as you already found, that you may try using `PathMatchSpecEx`.

Spec: https://msdn.microsoft.com/en-us/library/bb773727.aspx

I can pull this PR and try and play with it on my Windows dev machine when i get home.

**Comment by awemany** [comment]
@Justaphf : That would be awesome! The latest push to this PR, as above, compiles on windows as well and uses PathMatchSpec as per specification. I added a couple unit tests to make sure it works as intended - and they fail for the empty string in Windows. Maybe that's the only change necessary - to check the empty string as a special case.

**Comment by Justaphf** [comment]
@awemany I'm still working on testing, I got sidetracked trying to figure out why `make check` wasn't working natively on Windows using mingw.  I've got the why figured out, though haven't come up with a proper solution yet (I do have a work-around though).

Now that I've got `make check` working on my Windows dev machine, it looks like the new unit tests you added all pass, including the empty string case.  I think it must be an issue with Travis on the cross-compiled builds.  On my machine the results of the util unit tests are as follows:

```
Running 225 test cases...
2017-06-01 05:34:21 test/util_tests.cpp(61): Debug Assertion failed: "0"

*** No errors detected
PASS: test/test_bitcoin.exe
FAIL: qt/test/test_bitcoin-qt.exe
```
Not sure why `test_bitcoin-qt.exe` is failing for me, but it doesn't seem related to this PR.

I do think some additional unit tests should be added for robust full coverage of how the wildcards are used to ensure there is consistency between `fnmatch` and `PathMatchSpec`.  It looks like `fnmatch` supports some additional cases which `PathMatchSpec` doesn't ([seq] and [!seq]).  I haven't tested to verify yet, but depending on the search pattern this could lead to divergent results.  Additionally, I wanted to test against the original functionality to see what differences are there.  Based on comments it looks like there is at least 1 functional change, though I think it is an improvement (original made an assumption about not having consecutive *).

Will try to finish up testing tomorrow after work.

NOTE: Testing on 64-bit Windows 10 machine with 64-bit build of bitcoin.

**Comment by Justaphf** [comment]
@awemany  I've completed a large round of testing.  Not sure that `PathMatchSpec` is going to be a workable solution, I found several cases where the behavior diverges from `fnmatch`.

1. `PathMatchSpec` seems to be treating the `?` wildcard as "zero or one", while `fnmatch` is treating it as "exactly one".
2. In the case of consecutive `?`, `PathMatchSpec` is treating it as "n-1 or n", while `fnmatch` is treating it as "exactly n", where n is the number of consecutive `?`
3. `PathMatchSpec` treats an empty match pattern as "match anything", while `fnmatch` treats the empty match pattern as "only match an empty string".
4. `fnmatch` supports square brackets for escaping wildcards, `PathMatchSpec` treats the square bracket as a character literal to match.
5. 'PathMatchSpec' doesn't seem to collapse consecutive `*` wildcards whereas `fnmatch` does.

Specific test cases that were failing:
```
Line	Windows	Linux	Test
----	-------	-----	--------------------------------------
543	MATCH	FAIL	wildmatch("", "123")
551	FAIL	MATCH	wildmatch("1[?]3", "1?3")
555	FAIL	MATCH	wildmatch("1[*]3", "1*3")
567	MATCH	FAIL	wildmatch("?", "")
587	MATCH	FAIL	wildmatch("???", "12")
593	FAIL	FAIL	wildmatch("201?-??-??", "2017-06-01")
621	FAIL	MATCH	wildmatch("***", "")
643	MATCH	FAIL	wildmatch("20*-*-??", "20--x")
```
If you want to see all of the test cases I added, look at c985b91.  Some of the comments there are bogus as i was copy & pasting everywhere and sometimes I updated them and sometimes not.

Additionally, there seems to be an issue with testing cross-compiled on Travis because both the tests that only fail on Linux and the tests that only fail on native Windows are failing in the cross-compile, plus the empty string matched to empty string case.  So it's the worst of both worlds there.

**Comment by awemany** [comment]
@Justaphf : Thanks *a lot* for this detailed investigation!

@gandrewstone: With that info, what do you think the path forward should be here?

I see these options:

(1)- remove the functionality (which you said is not preferable)
(2)- leave it inconsistent between windows/unix (appears very ugly and hard to support)
(3)- use a regexp library and recode ?, *- wildcard strings into a regexp (if using boost, would be another library dependency)
(4)- use a suitable fnmatch(..) function from somewhere else

Regarding (3), this doesn't look to bad as boost has a regexp library which would probably does not make dependencies worse that much?
Regarding (4), BSD-licensed should be ok, right? So this http://web.mit.edu/freebsd/csup/fnmatch.c should be acceptable?

**Comment by gandrewstone** [comment]
let's do option 4.  BSD licensed is ok.  regexp has serious usability problems when used casually like on command lines...

**Comment by gandrewstone** [comment]
@awemany I think this is still WIP, right?  I've edited the title.  Please remove the WIP designator or comment when the PR is ready.

**Comment by awemany** [comment]
@gandrewstone: I noticed that the above BSD code uses the old 4-clause license which AFAIK is incompatible with MIT licensed code and would add in any case add the very annoying advertising burden. 

I looked around but don't find and appropriately licensed fnmatch.c around. What would be your plan of action now?

**Comment by awemany** [comment]
Ok, to tie this loose end up, I now implemented a simple but maybe somewhat limited wildcard matching function.

This function does not do any probing for different matches, but that also means that it should hopefully not run into any path explosion problems.

I took the PR out of WIP as I think this is ready to be reviewed as soon as I can rebase it to avoid the extra merges. Note that it should also be back-ported to previous releases.

File: src/test/util_tests.cpp [hunk 1]
```diff
@@ -728,43 +728,63 @@ bool IsStringTrueBadArgTest(const char *arg1)
...
-        BOOST_CHECK(IsStringTrue("enable"));
-        BOOST_CHECK(IsStringTrue("1"));
-        BOOST_CHECK(IsStringTrue("on"));
-        BOOST_CHECK(!IsStringTrue("false"));
-        BOOST_CHECK(!IsStringTrue("disable"));
-        BOOST_CHECK(!IsStringTrue("0"));
-        BOOST_CHECK(!IsStringTrue("off"));
-        BOOST_CHECK(IsStringTrueBadArgTest("bad"));
-    }
+    using namespace Logging;
```
Comment (sickpig) [inline src/test/util_tests.cpp]: minor nit. this indentation seems unrelated to the addition itself, I wonder if this has been done automatically by clang linter.

**Review by sickpig — APPROVED** [review]

**Comment by gandrewstone** [comment]
I like your KISS approach.  By making * a "lazy" match (match the first) we do not have to worry about backtracking which is sufficient for our application.  However I'm worried about the 100 char limit because I want to use this same matcher for CTweaks.  But since we are not backtracking, it should be very easy to convert the recursion into a while loop.

Two other common mini-features that might be useful (but I will merge without them):
1.  In "*" compare more than the first char.  Given a switch to a while loop, it should be pretty easy to find the string between wildcards (we don't have the escaped wildcard problem -- you can assume that any ? or * is a wildcard), and then test that string rather then just the first char.  This would solve the "*foo" not matching "fillyfoo" problem.
2. use "\*\*" (2 asterisks)  to mean a "greedy" match.  For example a pattern of "pfx**foo" searches forward to find the LAST foo rather than the first foo so will match "pfxffffoo" or "pfx/filly/foo".  [Prefix] <anything> suffix, is a very common search pattern so would be useful.  I guess this is not quite linux glob notation, since ** in glob greedily matches multiple directories.  But since we don't have the directory concept this defintion of ** provides a similar greedy function.

I'll implement the above if you are busy... let me know...

**Comment by sickpig** [comment]
@awemany PR need to be rebased.

**Comment by sickpig** [comment]
@awemany ?

**Comment by awemany** [comment]
@sickpig, @gandrewstone : Ok, rebased and I updated it to be iterative and increased the limit to 1024 chars. Go ahead and further squash the commit if you want - or let me know if I should do so.

**Comment by awemany** [comment]
One of the builds failed with a "KeyboardInterrupt" exception in excessive.py?!?

**Comment by sickpig** [comment]
@awemany great! thanks for the updates. 
wrt travis failure: it is a known issue. 
going to re-kick the job

**Comment by gandrewstone** [comment]
great, thanks!  merged...
